File: ruby/003.rb
#Q: What is the largest prime factor of the number 600851475143?
#A: 6857

class Fixnum
  def prime?
    return false if self % 2 == 0
    3.step(self**0.5, 2){ |i| return false if self % i == 0 }
    return true
  end
end

num = 600851475143
fac = 1
3.step(num/2, 2){ |i|
  if i.prime? && num % i == 0
    fac *= i
    break i if fac == num
  end
}

File: zig/003.zig
// Q: What is the largest prime factor of the number 600851475143?
// A: 6857

const print = @import("std").debug.print;

fn isPrime(n: u64) bool {
    if (n % 2 == 0) return false;

    var i:u64 = 3;
    while (i * i < n) {
        if (n % i == 0) return false;
        i += 2;
    }

    return true;
}

pub fn main() void {
    var n:u64 = 600851475143;
    var fac:u64 = 1;

    var i:u64 = 3;
    while (i < n / 2) : (i += 2) {
        if (n % i == 0 and isPrime(i)) fac *= i;
        if (fac == n) break;
    }

    print("{}\n", .{i});
}
